Document how to cast / recast / change type of pointer inside Ptr docs

freemin7 opened the pull request:
It took me a while to figure out how to do this and wished it would be documented here or in a similar place.

freemin7 committed:
If someone hacks on pointers they likely want to know that

It took me a while to figure out how to do this and wished it would be documented here or in a similar place.

freemin7 committed:
strip access white space

freemin7 committed:
Document how to cast / recast / change type of pointer inside Ptr docs

freemin7 commented:
This is a pure change to documentation making it easier for people to mess or observe via Julias memory.
This is only possible through other unsafe methods in the same file. While using reinterpret makes perfect sense
to change the type of pointers i did not consider it but i checked the docs and didn't find it either. Finally i found an issue which had this information and wanted to save others the same hunt.

LilithHafner commented:
Thanks! Would you please format this as a `see also` section? You can `grep` for `"see also"` in the julia repo to see examples.

freemin7 commented:
I took the syntax from an See also section but decided to add additional information as the documentation of `reinterpret` does not have a subsection on pointers. I feel like downgrading this to a "See also section" would require adding a  subsection to `reinterpret` to contain the same information (albeit with one more look up).

Can you confirm that you prefer a See also section + a new subsection to `reinterpret` over the current solution?

LilithHafner commented:
1) is not ideal because conversion is much more widely used than `Ptr`, so offering this method to people who query `?convert` is less precise than offering it to people who query `?Ptr`. 3 suffers from a similar issue.

Perhaps a variant of 2 would be good: adding a sentence after the existing `Ptr` docstring rather than in the middle of it.

freemin7 committed:
Elaborate on role of T in pointer type

This information is based on a conversation in the Julia Slack internals channel.

freemin7 commented:
I added a more sections which elaborates on the role the type T plays in pointers. I think this information is best collected in a central place. It also makes the reinterpret sentence stand out less.

LilithHafner reviewed (commented):
Thanks, this is much better :)

vtjnash commented on base/pointer.jl:
```diff
@@ -3,8 +3,16 @@
+
+!!! warn
+    When T is an [`Union`](@ref) type, the pointer does not contain the necessary information
+    to reconstruct what elements in the array had which type. Such pointers can't be use to
```
s/immutable/mutable I assume

freemin7 commented on base/pointer.jl:
```diff
@@ -3,8 +3,16 @@
+A memory address referring to data of type `T`. However there is no guarantee that the
 memory is actually valid, or that it actually represents data of the specified type.
+Some operations like [`unsafe_load`](@ref) are only supported when T is an [`isbitstype`](@ref)
+while [`unsafe_wrap`](@ref) also supports abstract types.
```
It is not meant as an alternative to pointers. `unsafe_wrap` takes a pointer as an argument and if it is a Ptr{T} where T is an abstract type it will still worked, while `unsafe_load` will throw an error if given such a pointer.

freemin7 committed:
Grammar tweak

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

freemin7 committed:
remove excess space

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>

freemin7 committed:
modify warning header

Co-authored-by: Lilith Orion Hafner <lilithhafner@gmail.com>
